engine/uState.ml

```diff
@@ -59,28 +77,34 @@ let rec repr q m = match QMap.find q m with
 
 let set q qv m =
   let q = repr q m in
+  let q = match q with QVar q -> q | QProp | QSProp | QType -> assert false in
   let qv = match qv with QVar qv -> repr qv m | (QSProp | QProp | QType as qv) -> qv in
   match q, qv with
-  | QVar q, QVar qv ->
+  | q, QVar qv ->
     if Sorts.QVar.equal q qv then Some m
-    else Some (QMap.add q (Some (QVar qv)) m)
-  | QVar q, (QProp | QSProp | QType as qv) | (QProp | QSProp | QType as qv), QVar q ->
-    Some (QMap.add q (Some qv) m)
-  | (QSProp, QSProp) | (QProp, QProp) | (QType, QType) -> Some m
-  | (QSProp | QProp | QType), (QSProp | QProp | QType) ->
-    None
+    else Some { qmap = QMap.add q (Some (QVar qv)) m.qmap; above = QSet.remove q m.above }
+  | q, (QProp | QSProp | QType as qv) ->
+    if qv == QSProp && QSet.mem q m.above then None
```

Comment:
oh I missed this

Reply:
The code in the var-var case above is faulty though, we need to propagate the flag. I'll push a fix.